src/GL/Array.hs

```diff
@@ -175,34 +171,14 @@ load is = do
 
 
 bindArray :: (Has Check sig m, Has (Lift IO) sig m) => Array (v I) -> ArrayC v m a -> m a
-bindArray array (ArrayC m) = do
+bindArray array m = do
   bind (Just array)
-  a <- runReader array m
+  a <- runReader array (runLabelled m)
   a <$ bind @(Array _) Nothing
 
-class Monad m => HasArray v m | m -> v where
-  askArray :: m (Array (v I))
+type HasArray v sig m = HasLabelled Array (Reader (Array (v I))) sig m
 
+askArray :: HasLabelled Array (Reader (Array (v I))) sig m => m (Array (v I))
+askArray = runUnderLabel @_ @Array ask
 
-newtype ArrayC v m a = ArrayC { runArrayT :: ReaderC (Array (v I)) m a }
-  deriving (Applicative, Functor, Monad, MonadIO, MonadTrans)
-
-deriving instance HasArray     v   m => HasArray     v   (ProgramC u v o m)
-deriving instance HasArray     v   m => HasArray     v   (B.BufferC ty x m)
-deriving instance B.HasBuffer ty x m => B.HasBuffer ty x (ArrayC     v   m)
-deriving instance HasProgram u v o m => HasProgram u v o (ArrayC     v   m)
-
-instance HasArray v m => HasArray v (FreshC m) where
-  askArray = lift askArray
-
-instance HasArray v m => HasArray v (ReaderC r m) where
-  askArray = lift askArray
-
-instance HasArray v m => HasArray v (StateC s m) where
-  askArray = lift askArray
-
-instance Algebra sig m => Algebra sig (ArrayC v m) where
-  alg = ArrayC . send . handleCoercible
-
-instance Algebra sig m => HasArray v (ArrayC v m) where
-  askArray = ArrayC ask
```

Comment:
Using `HasLabelled` means that we can remove:

1. the `ArrayC` carrier type;
2. the `HasArray` typeclass;
3. its _n²_ instances;
4. and the instances for other similar functional dependency-bearing classes.

Not a bad payoff.